#### Chapter 5 Statements

##### 5.1 Simple Statements

statement = expression + semicolon

> Write a while loop that reads `cin` into `s` until `s` is `sought`.

`while (cin >> s && s != sought)`

##### 5.2 Statement Scope

#####5.3 Conditional Statements

> Is the following line legal?
>
> ```c++
> if (int i = read())
>     cout << i << endl;
> ```

Yes.

> Is the following `switch` statement legal?
>
> ```c++
> switch (n) {
>     case 0: int i; i = 0; cout << i << endl; break;
>     case 1: i = 1; cout << i << endl; break;
>     default: string s = "N/A";  cout << s << endl; break;
> }
> ```
> What if change `case 0` into
>
> ```c++
> case 0: int i = 0; cout << i << endl; break;
> ```

Yes. No.

##### 5.4 Iterative Statements

> What does the following range `for ` do?
>
> ```c++
> for (auto i : {5, 4, 3, 2, 1})
> 	cout << i << endl;
> ```

Count down from 5 to 1.

> What's the problem with the following `do-while` loop?
>
> ```c++
> int i = 5;
> do
>     cout << i-- << endl;
> while (i > 0)
> ```

Missing a semicolon after the parenthesized condition.

##### 5.5 Jump Statements

1. `break`
2. `continue`
3. `goto`
4. `return`

> Consider
>
> ```c++
> begin:
> int i = read();
> end:
> i = 42;
> ```
>
> Can we add an `goto end;` at the beginning?
>
> Can we add an `goto begin;` at the end?

No. Yes.

##### 5.6 `try` Blocks and Exception Handling

Exception handling:

1. detecting (`throw`)
2. handling (`try` and `catch`[^handler])

[^handler]: `catch` is called *exception handler*.

> What is the problem with the following program?
>
> ```c++
> int main()
> {
>     try {
>         cout << "Enter two integers: ";
>         int a, b;
>         cin >> a >> b;
>         if (b == 0)
>             throw runtime_error("Division by 0.");
>         cout << a << " / " << b << " = " << a / b << endl;
>     }
> 	return 0;
> }
> ```

`try` must be matched with `catch`, though `throw` may stand alone.

> Assume `err` is a `runtime_error`. How to print out its contents?

`cout << err.what() << endl;`

> Where is `runtime_error` defined?

`<stdexcept>`

> Point out the syntax error in
>
> ```c++
> for (unsigned short n; cout << "Enter an integer n: ", cin >> n; ) {
>     try
>         if (n > 20)
>             throw invalid_argument("Invalid argument: n must be in [0, 20].");
>         else
>             cout << n << "! = " << fact(n) << endl;
>     catch (invalid_argument err)
>         cout << err.what() << endl;
> }
> ```

Curly braces `{}` cannot be omitted following `try` and `catch ()`.